feg/gateway/services/n7_n40_proxy/notify/notify_handler.go

```diff
@@ -0,0 +1,206 @@
+// - Handles SmPolicyUpdateNotification from PCF convert the message to proto and sends it as
+//   PolicyReAuth to SessionProxyResponder(feg_relay)
+// - Handles SmPolicyTerminateNotification from PCF convert the message to prot and sends it as
+//   AbortSession to feg_relay.
+
+const (
+	NOTIFY_PATH      = "/"
+	UnableToDeliver  = 3002
+	EncodedSessionId = "encodedSessionId"
+	// Shutdown timeout is used as a timeout to exit the Echo Server Shutdown
+	ShutdownTimeout = 10 * time.Second
+)
+
+type NotificationHandler struct {
+	config        *n7.N7ClientConfig
+	notifyServer  *echo.Echo
+	cloudRegistry service_registry.GatewayRegistry
+}
+
+func NewNotificationHandler(
+	config *n7.N7ClientConfig,
+	cloudReg service_registry.GatewayRegistry,
+) *NotificationHandler {
+	return &NotificationHandler{
+		config:        config,
+		notifyServer:  echo.New(),
+		cloudRegistry: cloudReg,
+	}
+}
+
+// RegisterHandlers registers the UpdateNotification and TerminateNotification handlers.
+// The notification url is of the form {notifyRoot}/{encodedSessionId}/{operation}
+//   Example: http://magma-feg.magma.com/sm-policy-control/v1/MTIzNDU2Nzg5MDsxMjM0NQo=/update
+//   Here notifyRoot = http://magma-feg.magma.com/sm-policy-control/v1
+//        encodedSessionId = MTIzNDU2Nzg5MDsxMjM0NQo= (Session-Id is urlencoded)
+//        operation = update
+// This notification url is send to PCF in the SmPolicyCreate request
+func (handler *NotificationHandler) RegisterHandlers() error {
+	urlDef, err := url.ParseRequestURI(handler.config.NotifyApiRoot)
+	if err != nil {
+		return fmt.Errorf("error parsing notify api root - %s", err)
+	}
+	basePath := fmt.Sprintf("%s/:%s", urlDef.Path, EncodedSessionId)
+	// Register the handlers
+	handler.notifyServer.POST(basePath+"/update", handler.postSmPolicyUpdateNotification)
+	handler.notifyServer.POST(basePath+"/terminate", handler.postSmPolicyTerminateNotification)
+	return nil
+}
+
+func (handler *NotificationHandler) Start() (string, error) {
+	errChan := make(chan error)
+	addr, err := net.ResolveTCPAddr("tcp", handler.config.LocalAddr)
+	if err != nil {
+		return "", err
+	}
+
+	go func() {
+		err = handler.notifyServer.Start(addr.String())
+		if err != nil {
+			errChan <- err
+		}
+	}()
+
+	return sbi.WaitForSbiServerAndGetAddr(handler.notifyServer, errChan)
+}
+
+func (handler *NotificationHandler) Shutdown() {
+	ctx, cancel := context.WithTimeout(context.Background(), ShutdownTimeout)
+	defer cancel()
+	handler.notifyServer.Shutdown(ctx)
+}
+
+// postSmPolicyUpdateNotification handles SM policy update notification request from PCF.
+func (handler *NotificationHandler) postSmPolicyUpdateNotification(ctx echo.Context) error {
+	client, err := relay.GetSessionProxyResponderClient(handler.cloudRegistry)
+	if err != nil {
+		glog.Errorf("postSmPolicyUpdateNotification failed to get SessionProxyResponderClient: %s", err)
+		return fmt.Errorf("internal server error")
```

Comment:
what is the logic on not returning the error string here?
Will the PCF complain?

Note i mean all the returns in here

Reply:
Not sure how this is implemented, but pre my understanding we have http error code that PCF should be using to understand the type of error. It is true that PCF doesn't need to know about our implementation. But why do we care on the specific message?